src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs

```diff
@@ -310,9 +391,34 @@ private static bool RequiresRefOrOut(BindValueKind kind)
 
 #nullable enable
 
+        private AccessorKind GetIndexerAccessorKind(BoundIndexerAccess indexerAccess, BindValueKind valueKind)
+        {
+            var coreValueKind = valueKind & ValueKindSignificantBitsMask;
+            var returnsByRef = indexerAccess.Indexer.RefKind != RefKind.None;
+            AccessorKind kind = (returnsByRef, coreValueKind) switch
+            {
+                (true, _) => AccessorKind.Get,
+                (false, BindValueKind.CompoundAssignment) => AccessorKind.Both,
+                (false, BindValueKind.Assignable) => AccessorKind.Set,
+                (false, BindValueKind.RValue) => AccessorKind.Get,
+
+                // These combinations of not returns by ref but having a ref accessor kind 
+                // occur in error scenarios.
+                (false, BindValueKind.RefOrOut) => AccessorKind.Get,
+                (false, BindValueKind.RefAssignable) => AccessorKind.Get,
+                (false, BindValueKind.ReadonlyRef) => AccessorKind.Get,
+                _ => AccessorKind.Unknown,
+            };
+            Debug.Assert(kind != AccessorKind.Unknown);
+            return kind;
+        }
+
         private BoundIndexerAccess BindIndexerDefaultArgumentsAndParamsCollection(BoundIndexerAccess indexerAccess, BindValueKind valueKind, BindingDiagnosticBag diagnostics)
         {
-            var useSetAccessor = valueKind == BindValueKind.Assignable && !indexerAccess.Indexer.ReturnsByRef;
+            var coreValueKind = valueKind & ValueKindSignificantBitsMask;
+            var returnsByRef = indexerAccess.Indexer.RefKind != RefKind.None;
```

Comment:
Previously, we were using `indexerAccess.Indexer.ReturnsByRef` which appears to check `RefKind == RefKind.Ref` only. Was there a bug when the indexer return was `ref readonly`?

Reply:
The `== BindValueKind.Assignable` seemed suspicious all on its own as `valueKind` is a bit flag structure. Agree that my current logic of not thinking about `ReadOnlyRef` seems wrong. Will undo that.